pkg/storage/ingest/reader.go

```diff
@@ -80,10 +80,15 @@ type PartitionReader struct {
 	consumerGroup                         string
 	concurrentFetchersMinBytesMaxWaitTime time.Duration
 
-	// fetchingMtx protects both the client and the fetcher but never at the same time.
-	fetchingMtx sync.Mutex
-	client      *kgo.Client
-	fetcher     fetcher
+	// client and fetcher are both start after PartitionReader creation. Fetcher could also be
+	// replaced during PartitionReader lifetime. To avoid concurrency issues with functions
+	// getting their pointers (e.g. BufferedRecords()) we use atomic / mutex to protect.
+	client atomic.Pointer[kgo.Client]
+
+	// We use a mutex for fetcher because it's an interface and we have to check against the
+	// nil case too.
+	fetcherMx sync.Mutex
+	fetcher   fetcher
```

Comment:
nitpick: can't we use a `atomic.Pointer[*fetcher]`? `*fetcher` will never be `nil`, but the `fetcher` it's pointing to might sometimes be